### ISSUE TITLE
C# style fomods fail with "The method or operation is not implemeted" & errorcode 5

### ISSUE THREAD

**Issue by GrantSP** [Modorganizer2/modorganizer#243]
# The problem
MO2 cannot open C# style fomod mods with the errors mentioned above

# Environment

* Mod Organizer Version that exhibits the issue: 2.1.1 revision 0a4fa10
* Last Mod Organizer Verison that did not exhibit the issue (if applicable): unsure
* Desktop OS/version used to run Mod Organizer: Windows 10 1709

# Details

If necessary, describe the problem you have been experiencing in more detail.

# Link to Mod Organizer logs

## USVFS
```
14:31:34.298 [I] usvfs dll 0.3.1.0-beta5 initialized in process 9828
14:31:34.298 [D] remove from process 9828
14:31:34.298 [I] create config in 9828
14:31:34.298 [I] 0 processes - 0
14:31:34.298 [I] mod_organizer_instance_1 created in process 9828
14:31:34.298 [I] attached to mod_organizer_instance_1 with 1 nodes, size 65536
14:31:34.298 [I] inv_mod_organizer_instance_1 created in process 9828
14:31:34.298 [I] attached to inv_mod_organizer_instance_1 with 1 nodes, size 65536
14:31:34.298 [D] context current shm: mod_organizer_instance (now 2 connections)

```
## MO Interface
https://gist.github.com/GrantSP/fc48cafd21ff74274502156572386b80

**Comment by SuperSandro2000** [comment]
Project Nevada uses such installer I think.

**Comment by SuperSandro2000** [comment]
Note: Is working on latest Nexus Release 2.1.1

**Comment by SuperSandro2000** [comment]
This https://github.com/Nexus-Mods/Nexus-Mod-Manager/commit/03448e0eb02e08f37d7b66507d0537ab67841321 broke it. For now use this files or grab them from the latest Nexus Release. I don't know if we just revert to an earlier commit or fix the api.
[NCC.zip](https://github.com/LePresidente/modorganizer/files/1763675/NCC.zip)

**Comment by GrantSP** [comment]
Latest dev build (Dev build 0.1 24/02/2018) appears to have fixed this issue.
Curiously some users noted that they were able to get some C# fomods to work before hand, just most were failing. I have no idea why that would be the case unless there is specific function in the script that is breaking the execution. It should be noted however that no C# fomods worked for me, even those that were said to work for others.
It would be great to get this functionality nailed down to code that doesn't rely on the NCC installer.

**Comment by GrantSP** [comment]
Update.
Apparently this latest fix does not work for all C# fomods. Project Nevada's main mod install fails during the file copy/install phase after correctly presenting the actual fomod selection screen. Oddly PN - Extra Options, which uses a very similar fomod, progresses perfectly.

**Comment by SuperSandro2000** [comment]
I tested it with MO1 NCC, installer_ncc and installer_fomod in MO2 and it did not work like in MO1.
I suspect that the issue is somewhere else unrelated to NCC. Needs further investigating.

**Comment by Silarn** [comment]
Solution will involve preextracting the fomod archive and then just running the script on that. The extraction step is what causes slowdowns and hangs.

**Comment by GrantSP** [comment]
Somewhat related to this, but possibly a whole level of extra coding pain, is the way some archives are packaging *.fomod files inside a folder and MO doesn't know what to do with it.
e.g. [Interior Lighting Overhaul for New Vegas](https://www.nexusmods.com/newvegas/mods/35794/)
The only way MO can install this mod properly is to first extract the *.fomod and then place that in downloads.

**Comment by SuperSandro2000** [comment]
@GrantSP I think this is a problem with the devs as fomods are already compressed and can be uploaded to Nexus. I would take the non mo version for ILO. If you really think this is an issue I think it would be a better idea to open a new issue because this would be technical an enhancement and not a bug.

**Comment by Al12rs** [comment]
Silarn published dev build 0.3 containing a fix for this problem on the discord server under #builds . 
Please test it and report if the problem does is indeed fixed.

**Comment by Silarn** [comment]
I ended up rewriting a bit of the NCC file handling code. There must have been some sort of change between Qt in 1.x versions and the latest Qt versions we're using now that cause problems with threading. Like the resizing window but when 7z is extracting media files, the NCC extraction process would also get hung because the NMM code tries to thread the process out the wazoo.

They implemented a thread safe wrapper class to try to make this work, but for some files this just causes it to hang with a dangling thread. I'm not sure why it fails with NCC if the NMM code works fine, but having the threaded component simply collect the list of files needed to extract and then performing the extraction and installation process in the final method using this list allows the extraction to work without any hangs and should be about as fast as it used to be without having to extract the entire archive.
